python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -896,107 +871,120 @@ class PyInferenceRequest
     : public PyWrapper<struct TRITONSERVER_InferenceRequest> {
  public:
   DESTRUCTOR_WITH_LOG(PyInferenceRequest, TRITONSERVER_InferenceRequestDelete);
-
-  using ReleaseFn = std::function<void(
-      std::shared_ptr<PyInferenceRequest>, uint32_t, py::object)>;
-
   // Defer definition until PyServer is defined
   PyInferenceRequest(
       PyServer& server, const std::string& model_name,
       const int64_t model_version);
-
   explicit PyInferenceRequest(
       struct TRITONSERVER_InferenceRequest* r, const bool owned)
       : PyWrapper(r, owned)
   {
   }
 
-
-  // Use internally when interacting with C APIs that takes ownership,
-  // this function will also release the ownership of the callback resource
-  // because once the ownership is transferred, the callback resource
-  // will be accessed in the callback pipeline and should not be tied to the
-  // PyWrapper's lifecycle. The callback resource will be released in the
-  // Triton C callback wrapper.
-  struct TRITONSERVER_InferenceRequest* Release()
+  void SetReleaseCallback(const std::shared_ptr<PyInferenceRequest>& request)
   {
-    // Note that Release() doesn't change ownership as the
-    // same PyInferenceRequest will be passed along the life cycle.
-    allocator_callback_resource_.release();
-    response_callback_resource_.release();
-    return triton_object_;
+    // The request wrapper needs to be kept alive until the release callback is
+    // invoked, so the Triton request can be deleted after the core/backend is
+    // done with it.
+    std::unique_ptr<std::shared_ptr<PyInferenceRequest>> request_ptr(
+        new std::shared_ptr<PyInferenceRequest>(request));
+    ThrowIfError(TRITONSERVER_InferenceRequestSetReleaseCallback(
+        triton_object_, ReleaseCallback,
+        reinterpret_cast<void*>(
+            request_ptr.release()) /* request_release_userp */));
+  }
+  static void ReleaseCallback(
+      struct TRITONSERVER_InferenceRequest* request, const uint32_t flags,
+      void* userp)
+  {
+    std::unique_ptr<std::shared_ptr<PyInferenceRequest>> request_ptr(
```

Comment:
I almost prefer to be more verbose here, 
```
std::shared_ptr<PyInferenceRequest> * request_ptr = reinterpret_cast<std::shared_ptr<PyInferenceRequest>*>(userp));
delete request_ptr;
```

Reply:
@rmccorm4, @GuanLuo for style - I see the general goodness of unique_ptr - but it kind of obscures what's happening here ....

Reply:
yes, removed the unique_ptr wrapper: [Remove unique_ptr wrapper on release callback](https://github.com/triton-inference-server/core/pull/426/commits/141483fa946cf5aa2610b6ea08facc3cc89bc705)